core/lib/zksync_core/src/api_server/tx_sender/proxy.rs

```diff
@@ -25,20 +113,41 @@ impl TxProxy {
         let client = HttpClientBuilder::default().build(main_node_url).unwrap();
         Self {
             client,
-            tx_cache: RwLock::new(HashMap::new()),
+            tx_cache: TxCache::default(),
         }
     }
 
     pub async fn find_tx(&self, tx_hash: H256) -> Option<L2Tx> {
-        self.tx_cache.read().await.get(&tx_hash).cloned()
+        self.tx_cache.get_tx(tx_hash).await
     }
 
     pub async fn forget_tx(&self, tx_hash: H256) {
-        self.tx_cache.write().await.remove(&tx_hash);
+        self.tx_cache.remove_tx(tx_hash).await;
+    }
+
+    pub async fn save_tx(&self, tx: L2Tx) {
+        self.tx_cache.push(tx).await;
+    }
+
+    pub async fn get_nonces_by_account(&self, account_address: Address) -> Vec<Nonce> {
+        self.tx_cache.get_nonces_for_account(account_address).await
     }
 
-    pub async fn save_tx(&self, tx_hash: H256, tx: L2Tx) {
-        self.tx_cache.write().await.insert(tx_hash, tx);
+    pub async fn next_nonce_by_initiator_account(
+        &self,
+        account_address: Address,
+        current_nonce: u32,
+    ) -> Nonce {
+        let mut expected_nonce = current_nonce.saturating_sub(1);
+        let nonces = self.get_nonces_by_account(account_address).await;
```

Comment:
As discussed earlier, we may have `nonces` less than `current_nonce` because they are not cleaned up yet; these should be ignored. IIUC, in the current approach not ignoring them can lead to bogus value returned; e.g., `current_nonce == 2` and `nonces == [1, 2, 3]` should output 4, but will output 2. To ignore nonces efficiently, you can use [`BTreeSet::range()`](https://doc.rust-lang.org/std/collections/struct.BTreeSet.html#method.range).